### PR TITLE
Create a PD volume when deploying the federation with federation-up.

### LINKED ISSUE #28248: Back federation etcd with persistent storage
As mentioned in kubernetes/kubernetes.github.io#662, the federation etcd will lose its state any time the pod dies, unless we modify `federation/manifests/federation-apiserver-deployment.yaml` to add a PVC.

We'll also need to create a PV *and* modify the turnup script to create the backing volume in the cloud provider.

@kubernetes/sig-cluster-federation

### PR DESCRIPTION
**Pull request by mml**
Tested only on GCE, where it worked.

Fixes #28248.

### CODE REVIEW

**Comment by mml** [comment]
I will squash.

**Comment by mml** [comment]
```
  SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 324 Skipped PASS
```

File: federation/cluster/common.sh [hunk 1]
```diff
@@ -45,7 +45,7 @@ if [[ -z "${FEDERATION_PUSH_REPO_BASE}" ]]; then
     fi
 fi
 
-FEDERATION_IMAGE_REPO_BASE=${FEDERATION_IMAGE_REPO_BASE:-'gcr.io/google_containers'}
+FEDERATION_IMAGE_REPO_BASE=${FEDERATION_IMAGE_REPO_BASE:-gcr.io/google_containers}
```
Comment (quinton-hoole) [inline federation/cluster/common.sh]: Is this intentional?  Seems unrelated to this fix?

File: federation/cluster/common.sh [hunk 2]
```diff
@@ -87,7 +87,18 @@ function create-federation-api-objects {
       fi
     fi
 
-    export FEDERATION_SERVICE_CIDR=${FEDERATION_SERVICE_CIDR:-"10.10.0.0/24"}
+    if [[ "${KUBERNETES_PROVIDER:-}" == "gce" || "${KUBERNETES_PROVIDER:-}" == "gke" ]]; then
+      PD_PLUGIN_NAME='kubernetes.io/gce-pid'
+    elif [[ "${KUBERNETES_PROVIDER:-}" == "aws" ]]; then
+      PD_PLUGIN_NAME='kubernetes.io/aws-ebs'
+    else
+      echo "not sure how to provision PD volume for ${KUBERNETES_PROVIDER:-}"
```
Comment (quinton-hoole) [inline federation/cluster/common.sh]: I would error out here, rather than just print a warning.

File: federation/cluster/common.sh [hunk 3]
```diff
@@ -87,7 +87,18 @@ function create-federation-api-objects {
       fi
     fi
 
-    export FEDERATION_SERVICE_CIDR=${FEDERATION_SERVICE_CIDR:-"10.10.0.0/24"}
+    if [[ "${KUBERNETES_PROVIDER:-}" == "gce" || "${KUBERNETES_PROVIDER:-}" == "gke" ]]; then
+      PD_PLUGIN_NAME='kubernetes.io/gce-pid'
```
Comment (quinton-hoole) [inline federation/cluster/common.sh]: Surely this happens automatically on these providers?  Why would I ever want an aws-ebs on gce?  This is more a PVC question than a criticism of this PR.  I was not aware that PVC's needed to be told what provider to use. I would have thought that they would default to the native PV plugin on the provider.  @saad-ali @thockin ?

File: federation/manifests/federation-etcd-pvc.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: {{.FEDERATION_APISERVER_DEPLOYMENT_NAME}}-etcd-claim
+  annotations:
+    volume.alpha.kubernetes.io/storage-class: "{{.PD_PLUGIN_NAME}}"
```
Comment (quinton-hoole) [inline federation/manifests/federation-etcd-pvc.yaml]: Aah, now I understand.  I would have guessed that there would be a default for this per cloud provider, but I guess it's still in alpha, so perhaps the answer is "not yet".

**Comment by quinton-hoole** [comment]
@mml Only minor nits.  Feel free to apply the label once you've addressed them to your liking.

File: federation/cluster/common.sh (hunk 3 above)
Comment (thockin) [inline federation/cluster/common.sh]: I assume you wanted "pd" not "pid".  Your UNIX is showing.  But you don't need to say anything special here.  Any value will do, for now - that's what makes it Alpha.  It only works on AWS, GCE, and OpenStack, and it knows what kind of volume to provision based on that.

File: federation/manifests/federation-etcd-pvc.yaml (hunk 4 above)
Comment (thockin) [inline federation/manifests/federation-etcd-pvc.yaml]: @quinton-hoole you're right - literally any value will do here.

File: federation/cluster/common.sh (hunk 1 above)
Comment (mml) [inline federation/cluster/common.sh]: It is unrelated, but my text editor goes bananas with the quotes in place.  I'll drop them from this.

**Comment by mml** [comment]
Grrr... now what?

```console
% cluster/kubectl.sh create -f crap.yaml
unable to decode "crap.yaml": [pos 121]: json: expect char '"' but got char 't'
```

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: federation-apiserver-etcd-claim
  annotations:
    volume.alpha.kubernetes.io/storage-class: yes
  namespace: federation
  labels:
    app: federated-cluster
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
```

**Comment by thockin** [comment]
you gonna love this.  YAML treats "yes" as boolan(true).  Try "yes" with
quotes :)

File: federation/cluster/common.sh (hunk 3 above)
Comment (saad-ali) [inline federation/cluster/common.sh]: @quinton-hoole In the alpha version of dynamic provisioning, provisioners are hard coded based on cloud provider (see [this](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/plugins.go#L110)). Moving forward, however, there will not necessarily be a 1:1 coupling between cloud provider and provisioner, see https://github.com/kubernetes/kubernetes/pull/26908

CC @pmorie
